### ISSUE TITLE
Primivite's vjp not implemented

### ISSUE THREAD

**Issue by Krupskis** [ml-explore/mlx#267]
Can I ask which primitive of these does not have a vjp yet?

```
def evaluate_loss(model, X, Y):
    return mx.mean(nn.losses.cross_entropy(model(X), Y))


def eval_fn(X, Y):
    return mx.mean(mx.argmax(X, axis=1) == Y)


def max_pool2d(x, kernel_size=(2, 2), stride=None):
    kh, kw = kernel_size
    dh, dw = stride if stride else kernel_size
    batch, height, width, channels = x.shape

    new_height = (height - kh) // dh + 1
    new_width = (width - kw) // dw + 1
    pooled = mx.zeros((batch, new_height, new_width, channels))
    for i in range(0, height - kh + 1, dh):
        for j in range(0, width - kw + 1, dw):
            window = x[:, i : i + kh, j : j + kw, :]
            pooled[:, i // dh, j // dw, :] = mx.max(window)

    return pooled


class TinyConvNet(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=(3, 3))
        self.conv2 = nn.Conv2d(32, 64, kernel_size=(3, 3))
        self.ln1 = nn.LayerNorm(32)
        self.ln2 = nn.LayerNorm(64)
        self.linear = nn.Linear(5 * 5 * 64, 10)

    def __call__(self, x):
        x = x.reshape(-1, 28, 28, 1)
        out = self.conv1(x)
        out = nn.relu(out)
        out = max_pool2d(out)
        out = self.ln1(out)
        out = self.conv2(out)
        out = nn.relu(out)
        out = max_pool2d(out)
        out = self.ln2(out)
        out = out.reshape(out.shape[0], -1)
        out = self.linear(out)
        return out


if __name__ == "__main__":
    batch_size = 256
    X_train, Y_train, X_test, Y_test = map(mx.array, fetch_mnist())
    model = TinyConvNet()

    optimizer = optim.SGD(learning_rate=1e-1)
    loss_and_grad_fn = nn.value_and_grad(model, evaluate_loss)
    count = 0
    for _ in range(500):
        ix = mx.random.randint(0, X_train.shape[0], [batch_size])

        Xb = X_train[ix]
        Yb = Y_train[ix]

        loss, grads = loss_and_grad_fn(model, Xb, Yb)
        print("loss", loss.shape)
        optimizer.update(model, grads)
```

**Comment by awni** [comment]
Yes it's this line, our scatter needs a VJP.

```
            pooled[:, i // dh, j // dw, :] = mx.max(window)
```

Your max pool 2d also will be extremely slow, MLX is not designed to efficiently handle that type of iterative indexing.

A much better/faster bet for you is to do some combindation of `mx.as_strided` + `mx.max(.., axis=..)`, but it might take a little indexing-fu to get all the shapes / strides right.

```
as_strided(...) method of builtins.PyCapsule instance
    as_strided(a: array, /, shape: Optional[List[int]] = None, strides: Optional[List[int]] = None, offset: int = 0, *, stream: Union[None, Stream, Device] = None) -> array
    
    Create a view into the array with the given shape and strides.
    
    The resulting array will always be as if the provided array was row
    contiguous regardless of the provided arrays storage order and current
    strides.
    
    .. note::
       Note that this function should be used with caution as it changes
       the shape and strides of the array directly. This can lead to the
       resulting array pointing to invalid memory locations which can
       result into crashes.
    
    Args:
      a (array): Input array
      shape (list(int), optional): The shape of the resulting array. If
        None it defaults to ``a.shape()``.
      strides (list(int), optional): The strides of the resulting array. If
        None it defaults to the reverse exclusive cumulative product of
        ``a.shape()``.
      offset (int): Skip that many elements from the beginning of the input
        array.
    
    Returns:
      array: The output array which is the strided view of the input.
```

**Comment by awni** [comment]
I'm going to close this as a dup of #25  and #155 (both of which we need to add to nn and core respectively)!
